controllers/alertsconfig_controller.go

```diff
@@ -50,14 +71,212 @@ type AlertsConfigReconciler struct {
+		return ctrl.Result{}, nil
+	}
+
+	// Convert list of alerts from status into Map for easy access since
+	// we need to compare alert hash from status to newly calculated hash to find the difference
+	alertHashMap := alertsConfig.Status.Alerts
+
+	// if len(configs) > 0
+	for _, config := range alertsConfig.Spec.Alerts {
+		// Calculate checksum and compare it with the status checksum
+		exist, reqChecksum := utils.CalculateAlertConfigChecksum(ctx, config)
+		// if request and status checksum matches then there is change in this specific alert config
+		if exist && alertHashMap[config.AlertName].LastChangeChecksum == reqChecksum {
+			log.Info("checksum is equal. skipping")
+			//skip it
+			continue
+		}
+		// if there is a diff
+		// Get Alert CR
+
+		var wfAlert alertmanagerv1alpha1.WavefrontAlert
+		wfAlertNamespacedName := types.NamespacedName{Namespace: req.Namespace, Name: config.AlertName}
+		if err := r.Get(ctx, wfAlertNamespacedName, &wfAlert); err != nil {
+			log.Error(err, "unable to get the wavefront alert details for the requested name", "wfAlertName", config.AlertName)
+
+			// Update the status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+		wfAlertBytes, err := json.Marshal(wfAlert.Spec)
+		if err != nil {
+			// update the status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+
+		// execute Golang Template
+		wfAlertTemplate, err := template.ProcessTemplate(ctx, string(wfAlertBytes), config.Params)
+		if err != nil {
+			//update the status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+		log.Info("Template process is successful", "here", wfAlertTemplate)
+
+		// Unmarshal back to wavefront alert
+		if err := json.Unmarshal([]byte(wfAlertTemplate), &wfAlert.Spec); err != nil {
+			// update the wfAlert status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+		// Convert to Alert
+		var alert wf.Alert
+
+		if err := wavefront.ConvertAlertCRToWavefrontRequest(ctx, wfAlert.Spec, &alert); err != nil {
+			errMsg := "unable to convert the wavefront spec to Alert API request. will not be retried"
+			log.Error(err, errMsg)
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+
+		// Create/Update Alert
+		if alertHashMap[config.AlertName].ID == "" {
+			// Create use case
+			if err := r.WavefrontClient.CreateAlert(ctx, &alert); err != nil {
+				r.Recorder.Event(&alertsConfig, v1.EventTypeWarning, err.Error(), "unable to create the alert")
+				state := alertmanagerv1alpha1.Error
+				if strings.Contains(err.Error(), "Exceeded limit setting") {
+					// For ex: error is "Exceeded limit setting: 100 alerts allowed per customer"
+					state = alertmanagerv1alpha1.Error
+				}
+				log.Error(err, "unable to create the alert")
+
```

Comment:
I think I have not updated to patch function here- I will update that and patch function basically requeueue it automatically if state is NOT error and in case of ClientExceededLimit we can skip retrying it since it is NOT error

Reply:
Okay, so you are updating the patch changes inside PatchIndividualAlertsConfigError